src/index.ts

```diff
@@ -0,0 +1,168 @@
+
+interface PluginConfig {
+  stages?: string[];
+  roleArn?: string;
+  filterPattern: string;
+  normalizedFilterID: boolean;
+  createLambdaPermission: boolean;
+  destinationARN: string;
+}
+
+const configDefaults = {
+  filterPattern: '',
+  normalizedFilterID: true,
+  createLambdaPermission: true,
+};
+
+interface CFObject<TProps> {
+  Type: string;
+  DependsOn?: string[];
+  Properties: TProps;
+}
+
+interface LambdaPermissionProps {
+  Action: string;
+  Principal: string;
+  FunctionName: string;
+}
+
+interface SubscriptionFilterProps {
+  DestinationArn: string;
+  FilterPattern: string;
+  LogGroupName: string;
+  RoleArn?: string;
+}
+
+type LambdaPermission = CFObject<LambdaPermissionProps>;
+type SubscriptionFilter = CFObject<SubscriptionFilterProps>;
+
+function getStage(options: ServerlessConfig, sls: ServerlessInstance) {
+  if (options.stage && options.stage !== '') {
+    return options.stage;
+  }
+  return sls.service.provider.stage;
+}
+
+class LogForwardingPlugin {
+  options: ServerlessConfig;
+
+  provider: AWSProvider;
+
+  serverless: ServerlessInstance;
+
+  config: PluginConfig | null = null;
+
+  hooks: object;
+
+  constructor(serverless: ServerlessInstance, options: ServerlessConfig) {
+    this.serverless = serverless;
+    this.options = options;
+    this.provider = this.serverless.getProvider('aws');
+    this.hooks = {
+      'package:initialize': this.updateResources.bind(this),
+    };
+  }
+
+  loadConfig() {
+    this.config = { ...configDefaults, ...this.serverless.service.custom.logForwarding };
+    if (this.config.destinationARN === undefined) {
+      throw new Error('Serverless-log-forwarding is not configured correctly. Please see README for proper setup.');
+    }
+  }
+
+  /**
+   * Updates CloudFormation resources with log forwarding
+   */
+  updateResources() {
+    this.loadConfig();
+    const stage = getStage(this.options, this.serverless);
+    if (this.config.stages && !this.config.stages.includes(stage)) {
+      this.serverless.cli.log(`Log Forwarding is ignored for ${stage} stage`);
+      return;
+    }
+    this.serverless.cli.log('Updating Log Forwarding Resources...');
+    const resourceObj = this.createResourcesObj();
+    const cfResources = this.getResources();
+    _.extend(cfResources, resourceObj);
+    this.serverless.cli.log('Log Forwarding Resources Updated');
+  }
+
+  private getResources() {
+    const { service } = this.serverless;
+    if (service.resources === undefined) {
+      service.resources = {
+        Resources: {},
+      };
+    }
+    if (service.resources.Resources === undefined) {
+      service.resources.Resources = {};
+    }
+    return service.resources.Resources;
+  }
+
+  createResourcesObj(): object {
+    const { service } = this.serverless;
+    const resourceObj = {};
+
+    const createLambdaPermission = this.config.createLambdaPermission && !this.config.roleArn;
+    const permissionId = 'LogForwardingLambdaPermission';
```

Comment:
Better to make `LogForwardingLambdaPermission` global variable

Reply:
addressed

Reply:
addressed